README.md
[projectUri]: https://github.com/sketch7/ngx.ux
[changeLog]: ./CHANGELOG.md
[releaseWorkflowWiki]: ./docs/RELEASE-WORKFLOW.md

[npm]: https://www.npmjs.com

# @ssv/ngx.ux
[![CI](https://github.com/sketch7/ngx.ux/actions/workflows/ci.yml/badge.svg)](https://github.com/sketch7/ngx.ux/actions/workflows/ci.yml)
[![npm version](https://badge.fury.io/js/%40ssv%2Fngx.ux.svg)](https://badge.fury.io/js/%40ssv%2Fngx.ux)

UX essentials for building apps, utilities which enables you writing richer apps easier.

**Quick links**

[Change logs][changeLog] | [Project Repository][projectUri]

## Installation

Get library via [npm]

```bash
npm install @ssv/ngx.ux
```

Choose the version corresponding to your Angular version:

 | Angular | library |
 | ------- | ------- |
 | 10+     | 2.x+    |
 | 4 to 9  | 1.x+    |

## Features
- Viewport (see below)
  - [Viewport Data](./src/viewport/viewport-data/README.md)

# Usage

## Register module

```ts
import { SsvUxModule } from "@ssv/ngx.ux";

@NgModule({
  imports: [
    SsvUxModule
  ]
}
export class AppModule {
}
```

## Viewport
Provides utilities to handle responsiveness easier based on the viewport (view size)

### Comparison Operands
| Operand | Description           |
| ------- | --------------------- |
| =       | Equals                |
| <>      | Not equals            |
| <       | Less than             |
| <=      | Less than or equal    |
| >       | Greater than          |
| >=      | Greater Than or equal |

<br>

### Size Types
These are the defaults, but they are configurable.

| Size Type | Size Range |
| --------- | ---------- |
| xsmall    | <=450      |
| small     | 451-767    |
| medium    | 768-992    |
| large     | 993-1200   |
| xlarge    | 1201-1500  |
| xxlarge   | 1501-1920  |
| xxlarge1  | >=1921     |


### Viewport Matcher Attribute (directive)
Structural directive which loads components based on a viewport sizing condition e.g. show ONLY if viewport is greater than xlarge.


#### Examples

```html
<!-- simple -->
<div *ssvViewportMatcher="'large'">
  show only when large
</div>

<!-- expression based - tuple (shorthand) *recommended usage* -->
<div *ssvViewportMatcher="['>=', 'xlarge']"> (see all operands and sizes)
  show when >= xlarge
</div>

<!-- expression based - object -->
<div *ssvViewportMatcher="{size: 'xlarge', operation: '<='}"> (see all operands and sizes)
  show when >= xlarge
</div>

<!-- includes -->
<div *ssvViewportMatcher="['large', 'xlarge']">
  show only when large, xlarge
</div>

<!-- excludes -->
<div *ssvViewportMatcher="''; exclude ['xsmall', 'small']">
  hide only when xsmall, small
</div>

<!-- match/else -->
<div *ssvViewportMatcher="['>=', 'xlarge']; else otherwise">
  show when >= xlarge
</div>

<ng-template #otherwise>
  show when expression is falsy (< xlarge)
</ng-template>

<!-- non structure syntax -->
<ng-template ssvViewportMatcher [ssvViewportMatcherExclude]="'xsmall'">
    (exclude xsmall)
</ng-template>
```

### Viewport Matcher Var (directive)
Structural directive which provides the condition var whether it matches or not (params are similar to `ssvViewportMatcher`).

```html
<!-- simple -->
<div *ssvViewportMatcherVar="let isLarge when 'large'">
  isLarge={{isLarge}}
</div>

<!-- expression based - tuple (shorthand) *recommended usage* -->
<div *ssvViewportMatcherVar="let isMediumDown when ['<=', 'medium']">
  isMediumDown={{isMediumDown}}
</div>

<!-- includes/or -->
<div *ssvViewportMatcherVar="let isLargeOrSmall when ['small', 'large']">
  isLargeOrSmall={{isLargeOrSmall}}
</div>
```

### Viewport Service

```ts
// get size type
this.viewport.sizeType$.pipe(
    tap(x => console.log("Viewport - sizeType changed", x)), // { type: 4, name: "xlarge", widthThreshold: 1500 }
  ).subscribe();
```

### Viewport for SSR
Since in SSR there is no way to know the client viewport size, we should at least determine device type and handle provide
3 different sizes based on device type e.g. `mobile`, `tablet` or `desktop` so the initial rendering will be closer based on device type.

The basic implementation allows to provide a device type `mobile`, `tablet` or `desktop` and there are static sizes for those.

```ts
import { UX_VIEWPORT_SSR_DEVICE } from "@ssv/ngx.ux";

const deviceType = deviceTypeFromServer;
{ provide: UX_VIEWPORT_SSR_DEVICE, useValue: deviceType },
```

The default implementation can also be replaced by implementing a small class as following:

```ts

export class SuperViewportServerSizeService {
  get(): ViewportSize {
    // do your magic..
    return size;
  }
}

import { ViewportServerSizeService } from "@ssv/ngx.ux";

@NgModule( {
  providers: [
    { provide: ViewportServerSizeService, useClass: SuperViewportServerSizeService }
  ]
}) export class AppModule {
}
```


## Configure
You can configure the existing resize polling speed and as well as provide your custom breakpoints.

### Custom Breakpoints
```ts
import { SsvUxModule, generateViewportSizeType } from "@ssv/ngx.ux";

const breakpoints = { // custom breakpoints - key/width
  smallest: 500,
  small: 700,
  medium: 1000,
  large: 1400,
  extralarge: 1600,
  xxlarge: 1800,
  fhd: 1920,
  uhd: 3840
};

  imports: [
    SsvUxModule.forRoot({
      viewport: {
        resizePollingSpeed: 66, // optional - defaults to 33
        breakpoints // provide the custom breakpoints
      }
    }),
  ],
```

### Override existing Breakpoints
```ts
import { SsvUxModule, UX_VIEWPORT_DEFAULT_BREAKPOINTS } from "@ssv/ngx.ux";

  imports: [
    SsvUxModule.forRoot({
      viewport: {
        breakpoints: {
          ...UX_VIEWPORT_DEFAULT_BREAKPOINTS, // use breakpoints provided with library
          xxlarge1: 2000, // override xxlarge1
          uhd: 3840 // add new breakpoint
        }
      }
    }),
  ],
```


## Getting Started

### Setup Machine for Development
Install/setup the following:

- NodeJS v18.16.0+
- Visual Studio Code or similar code editor
- TypeScript 5.0+
- Git + SourceTree, SmartGit or similar (optional)
- Ensure to install **global NPM modules** using the following:


```bash
npm install -g git gulp devtool
```


### Project Setup
The following process need to be executed in order to get started.

```bash
npm install
```


### Building the code

```bash
npm run build
```

### Running the tests

```bash
npm test
```

#### Watch
Handles compiling of changes.

```bash
npm start
```


#### Running Continuous Tests
Spawns test runner and keep watching for changes.

```bash
npm run tdd
```


### Preparation for Release

- Update changelogs
- bump version


Check out the [release workflow guide][releaseWorkflowWiki] in order to guide you creating a release and publishing it.

### Regen Examples
- `ng new examples --skip-install --create-application=false`
- `cd examples`
- `ng g example-app --routing --style=scss`
- `ng lint`
- `ng add @angular/material`

CHANGELOG.md
## [2.1.2](https://github.com/sketch7/ngx.ux/compare/2.1.1...2.1.2) (2023-06-15)

### Bug Fixes

- **version:** update `VERSION` correctly

## [2.1.1](https://github.com/sketch7/ngx.ux/compare/2.1.0...2.1.1) (2023-06-13)

### Features

- **viewport:** add `ComparisonOperationLiteral`
- **deps:** change peerDependency for rxjs to `>=6.0.0`

### Bug Fixes

- **viewport:** `ssvViewportMatcherVar` allow usage as `{size: 'large', operation: '<='}` with template type validation

## [2.1.0](https://github.com/sketch7/ngx.ux/compare/1.4.1...2.1.0) (2023-06-13)

### Features

- **deps:** update dependencies to support Angular v15
- **deps:** add dependency on `tslib: ^2.0.0`

### BREAKING CHANGES

- **viewport:** make several utils/models internal only
- **viewport:** change `ViewportSizeType` and `UX_VIEWPORT_DEFAULT_BREAKPOINTS` has been changed (see readme)
- **viewport:** remove `ViewportService.calculateItemsPerRow` function

## [1.4.1](https://github.com/sketch7/ngx.ux/compare/1.4.0...1.4.1) (2023-04-13)

### Features

- **viewport:** `ViewportService` - add `resizeSnap$`, `sizeTypeSnap$`, `sizeSnap$` which are unthrottled
- **viewport:** `ViewportDataService` `get$` add option to `throttle`

## [1.4.0](https://github.com/sketch7/ngx.ux/compare/1.3.0...1.4.0) (2020-11-25)

### Features

- **viewport:** implement `ssvViewportMatcherVar` directive

## [1.3.0](https://github.com/sketch7/ngx.ux/compare/1.2.2...1.3.0) (2020-11-11)

### Features

- **viewport:** implement viewport data - `ViewportDataService`, `ssvViewportData` pipe (see doc for more info)
- **viewport:** change `ViewportService.sizeTypeMap` - now maps by type (id) as well
- **viewport:** add `ViewportService.sizeTypeSnapshot`

### Refactor

- **module:** refactor main module config

## [1.2.2](https://github.com/sketch7/ngx.ux/compare/1.2.1...1.2.2) (2020-10-30)

### Features

- **viewport:** add `size$` to `ViewportService`

### Bug Fixes

- **viewport:** throw error when invalid size type is used in matcher

## [1.2.1](https://github.com/sketch7/ngx.ux/compare/1.2.0...1.2.1) (2020-10-28)

### Bug Fixes

- **build:** down version several dependencies packages in order to support angular 8 (and possibly less)
- **deps:** revert dependency on tslib `^1.10.0`

## [1.2.0](https://github.com/sketch7/ngx.ux/compare/1.1.1...1.2.0) (2020-10-27)

### Features

- **viewport:** implement configurable breakpoints (thanks to contrib [@joseph118](https://github.com/joseph118))
- **viewport:** add util method `generateViewportSizeType` to generate a representation of `ViewportSizeType`
- **deps:** now depends on tslib `^1.10.0`

### Refactor

- **all:** minor refactoring

### Chore

- **deps:** update dev dependencies
- **ci:** update node version
- **lint:** migrate from tslint to eslint

## [1.1.1](https://github.com/sketch7/ngx.ux/compare/1.1.0...1.1.1) (2020-05-29)

### Bug Fixes

- **viewport:** use `documentElement.clientWidth` to check viewport width on Safari only
- **viewport:** `getViewportSize` not being called on SSR

## [1.1.0](https://github.com/sketch7/ngx.ux/compare/1.0.0...1.1.0) (2019-08-22)

### Features

- **viewport:** add additional viewport breakpoint size `xxlarge1 >=1920`

## [1.0.0](https://github.com/sketch7/ngx.ux/compare/0.2.0...1.0.0) (2019-06-11)

### Chore

*No new fixes/features in this release*

- **deps:** update dev dependencies
- **ci:** update node version

## [0.2.0](https://github.com/sketch7/ngx.ux/compare/0.1.5...0.2.0) (2019-05-20)

- **viewport:** width now includes scrollbar to be more inline with css media queries e.g. `window.innerWidth` instead of `documentElement.clientWidth` (same for height)
- **viewport:** resize now emits more frequently, not only after stopping resizing for a duration of time e.g. use `auditTime` instead of `debounceTime`


### BREAKING CHANGES

- **viewport:** viewport size matching (including viewport matcher) size ranges now change according to the viewport size instead of the viewport container (e.g. scrollbar size is included)


## [0.1.5](https://github.com/sketch7/ngx.ux/compare/0.1.4...0.1.5) (2018-02-26)

### Features

- **viewport:** implemented `calculateItemsPerRow` in service


## [0.1.4](https://github.com/sketch7/ngx.ux/compare/0.1.3...0.1.4) (2018-02-21)

### Chore

update tooling


## [0.1.3](https://github.com/sketch7/ngx.ux/compare/0.1.2...0.1.3) (2018-01-31)

### Bug Fixes

- **viewport:** fix `sizeType$` was returning the initial `sizeType` on subscribe until resizing instead of the last value


## [0.1.2](https://github.com/sketch7/ngx.ux/compare/0.1.1...0.1.2) (2018-11-27)

### Bug Fixes

- **viewport:** fix `ssvViewportMatcher` when used on `ng-container` was trying to add cssClass and throwing an error


## [0.1.1](https://github.com/sketch7/ngx.ux/compare/0.1.0...0.1.1) (2018-11-26)

### Bug Fixes

- **viewport:** fix when using OnPush with `ssvViewportMatcher`


## [0.1.0](https://github.com/sketch7/ngx.ux) (2018-11-23)

Initial release

### Features

- **viewport:** viewport service implementation
- **viewport:** viewport size matcher implementation


src/viewport/viewport.model.ts
import { Dictionary } from "../internal/internal.model";
import { ViewportDataMatchStrategy } from "./viewport-data";

/**
 * The indices of each breakpoint provided based on the `UX_VIEWPORT_DEFAULT_BREAKPOINTS`.
 * @see UX_VIEWPORT_DEFAULT_BREAKPOINTS
 */
export enum ViewportSizeType {
	xsmall = 0,
	small = 1,
	medium = 2,
	large = 3,
	xlarge = 4,
	xxlarge = 5,
	xxlarge1 = 6,
}

export enum ComparisonOperation {
	equals = "=",
	notEquals = "<>",
	lessThan = "<",
	lessOrEqualThan = "<=",
	greaterThan = ">",
	greaterOrEqualThan = ">=",
}

export type ComparisonOperationKeyType = keyof typeof ComparisonOperation;
export type ComparisonOperationValueType = "=" | "<>" | "<" | "<=" | ">" | ">="; // todo: find a better way to do this
export type ComparisonOperationLiteral = ComparisonOperationKeyType | ComparisonOperationValueType;

export enum DeviceType {
	desktop = "desktop",
	mobile = "mobile",
	tablet = "tablet"
}

export interface UxViewportOptions {
	/** Polling speed on resizing (in milliseconds). e.g. the higher the number the longer it takes to recalculate. */
	resizePollingSpeed: number;

	/** Breakpoints to use. Key needs to match the size type and the value the width threshold.
	 * e.g. given width '1000' and `medium` is set to '992' => `large`.
	 */
	breakpoints: Dictionary<number>;

	/** Default data match strategy to use. */
	defaultDataMatchStrategy: ViewportDataMatchStrategy;
}

export interface ViewportSize {
	width: number;
	height: number;
}

export interface ViewportSizeTypeInfo {
	type: number;
	name: string;
	widthThreshold: number;
}

export interface ViewportMatchConditions {
	sizeType?: string | string[] | null;
	sizeTypeExclude?: string | string[] | null;
	expression?: ViewportSizeMatcherExpression;
}

export interface ViewportSizeMatcherExpression {
	size: string;
	operation: ComparisonOperationLiteral;
}
